cognite/neat/core/extractors/rules_to_graphql.py

```diff
@@ -0,0 +1,151 @@
+    invalid_names: set = get_invalid_names(transformation_rules.get_entity_names())
+
+    if invalid_names and not fix_names:
+        msg = "Entity names must only contain [_a-zA-Z0-9] characters and can start only with [_a-zA-Z]"
+        logging.error(f"{msg}, following entities {invalid_names} do not follow these rules!")
+        raise GraphQLError(f"{msg}, following entities {invalid_names} do not follow these rules!")
+    elif invalid_names and fix_names:
+        msg = "Entity names must only contain [_a-zA-Z0-9] characters and can start only with [_a-zA-Z]"
+        logging.warn(
+            f"{msg}, following entities {invalid_names} do not follow these rules! Attempting to repair names..."
+        )
+        warnings.warn(
+            f"{msg}, following entities {invalid_names} do not follow these rules! Attempting to repair names...",
+            stacklevel=2,
+        )
+
+    def _define_fields(property_definitions: pd.DataFrame) -> dict[str, GraphQLField]:
+        gql_field_definitions = {}
+        for property_, row in property_definitions.iterrows():
+            property_name = repair_name(property_, "property", fix_casing=fix_casing)  # type: ignore
+            # Node attribute
+            if row.property_type == "DatatypeProperty":
+                value_type_gql = DATA_TYPE_MAPPING[row.value_type]["GraphQL"]
+
+                # Case: Mandatory, single value
+                if row.min_count and row.max_count == 1:
+                    gql_field_definitions[property_name] = GraphQLField(GraphQLNonNull(value_type_gql))
+
+                # Case: Mandatory, multiple value
+                elif row.min_count and row.max_count != 1:
+                    gql_field_definitions[property_name] = GraphQLField(
+                        GraphQLNonNull(GraphQLList(GraphQLNonNull(value_type_gql)))
+                    )
+
+                # Case: Optional, single value
+                elif row.max_count == 1:
+                    gql_field_definitions[property_name] = GraphQLField(value_type_gql)
+
+                # Case: Optional, multiple value
+                else:
+                    gql_field_definitions[property_name] = GraphQLField(GraphQLList(value_type_gql))
+
+            # Node edge
+            else:
+                # Case: one to one edge
+                if row.min_count and row.max_count == 1:
+                    gql_field_definitions[property_name] = GraphQLField(
+                        gql_type_definitions[repair_name(row.value_type, "class", fix_casing=fix_casing)]
+                    )
+
+                # Case: one to many edge
+                else:
+                    gql_field_definitions[property_name] = GraphQLField(
+                        GraphQLList(gql_type_definitions[repair_name(row.value_type, "class", fix_casing=fix_casing)])
+                    )
+
+        return gql_field_definitions
+
+    for class_, properties in transformation_rules.to_dataframe().items():
```

Comment:
I suggest avoiding pandas in this case. You end up iterating each row, and then, I would suggest using the pydantic classes which gives you much more help by the IDE. In addition, this part of the code becomes independent of pandas, which is good if we decide to replace it in the future.

Reply:
Done! Also added additional validation of data model definitions.